lib/client/mfa/cli.go

```diff
@@ -0,0 +1,201 @@
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package mfa
+
+import (
+	"context"
+	"errors"
+	"fmt"
+	"io"
+	"sync"
+
+	"github.com/gravitational/trace"
+
+	"github.com/gravitational/teleport/api/client/proto"
+	"github.com/gravitational/teleport/api/utils/prompt"
+	wancli "github.com/gravitational/teleport/lib/auth/webauthncli"
+	wantypes "github.com/gravitational/teleport/lib/auth/webauthntypes"
+	"github.com/gravitational/teleport/lib/auth/webauthnwin"
+)
+
+// CLIPrompt is the default CLI mfa prompt.
+type CLIPrompt struct {
+	cfg    PromptConfig
+	writer io.Writer
+}
+
+// NewCLIPrompt returns a new CLI mfa prompt with the config and writer.
+func NewCLIPrompt(cfg *PromptConfig, writer io.Writer) *CLIPrompt {
+	return &CLIPrompt{
+		cfg:    *cfg,
+		writer: writer,
+	}
+}
+
+// Run prompts the user to complete an MFA authentication challenge.
+func (c *CLIPrompt) Run(ctx context.Context, chal *proto.MFAAuthenticateChallenge) (*proto.MFAAuthenticateResponse, error) {
+	runOpts, err := c.cfg.GetRunOptions(ctx, chal)
+	if err != nil {
+		return nil, trace.Wrap(err)
+	}
+
+	ctx, cancel := context.WithCancel(ctx)
+	defer cancel()
+	var wg sync.WaitGroup
+
+	type response struct {
+		kind string
+		resp *proto.MFAAuthenticateResponse
+		err  error
+	}
+	respC := make(chan response, 2)
+
+	// Use variables below to cancel OTP reads and make sure the goroutine exited.
+	otpCtx, otpCancel := context.WithCancel(ctx)
+	defer otpCancel()
+	otpDone := make(chan struct{})
+	otpCancelAndWait := func() {
+		otpCancel()
+		<-otpDone
+	}
+
+	// Fire TOTP goroutine.
+	if runOpts.promptTOTP {
+		wg.Add(1)
+		go func() {
+			defer wg.Done()
+			defer close(otpDone)
+
+			// Let Webauthn take the prompt below.
+			resp, err := c.promptTOTP(otpCtx, chal, true /*quiet*/)
+			respC <- response{kind: "TOTP", resp: resp, err: err}
+		}()
+	}
+
+	// Fire Webauthn goroutine.
+	if runOpts.promptWebauthn {
+		wg.Add(1)
+		go func() {
+			defer wg.Done()
+
+			// get webauthn prompt and wrap with otp context handler.
+			prompt := &webauthnPromptWithOTP{
+				LoginPrompt:      c.getWebauthnPrompt(ctx, runOpts.promptTOTP),
+				otpCancelAndWait: otpCancelAndWait,
+			}
+
+			resp, err := c.promptWebauthn(ctx, chal, prompt)
+			respC <- response{kind: "WEBAUTHN", resp: resp, err: err}
+		}()
+	}
+
+	go func() {
+		wg.Wait()
+		close(respC)
+	}()
+
+	for resp := range respC {
+		switch err := resp.err; {
+		case errors.Is(err, wancli.ErrUsingNonRegisteredDevice):
+			// Surface error immediately.
+		case err != nil:
+			c.cfg.Log.WithError(err).Debugf("%s authentication failed", resp.kind)
+			continue
+		}
```

Comment:
I know this is a copy-paste but it looks like this `errors.Is(err, wancli.ErrUsingNonRegisteredDevice)` is either useless, or there's missing code to actually surface this error immediately (cc @codingllama)

Reply:
A cursory search shows me this, which I vaguely remember as part of an UX improvements PR: https://github.com/gravitational/teleport/blob/master/lib/auth/webauthncli/fido2.go#L732


I think (hope?) that still works as intended.

Reply:
Oh I see, I was just forgetting how case statements in Go work. Spending too much time in my rust corner of the codebase, ignore me.

Reply:
It should still work. Basically when we get this error it ends the prompt immediately and returns the error. With other errors we just log it and wait for the otp prompt to complete.